drivers/wifi/eswifi/eswifi_core.c

```diff
@@ -0,0 +1,500 @@
+	}
+
+	/* Any IP assigned ? (dhcp offload or manually) */
+	err = __parse_ipv4_address(eswifi->buf, eswifi->sta.ssid,
+				   (u8_t *)&addr.s4_addr);
+	if (err) {
+		LOG_ERR("Unable to retrieve IP address");
+		goto error;
+	}
+
+	LOG_DBG("ip = %d.%d.%d.%d", addr.s4_addr[0], addr.s4_addr[1],
+		   addr.s4_addr[2], addr.s4_addr[3]);
+
+	net_if_ipv4_addr_add(eswifi->iface, &addr, NET_ADDR_DHCP, 0);
+
+	LOG_DBG("Connected!");
+
+	eswifi_unlock(eswifi);
+	return 0;
+
+error:
+	eswifi_unlock(eswifi);
+	return -EIO;
+}
+
+static int eswifi_disconnect(struct eswifi_dev *eswifi)
+{
+	char disconnect[] = "CD\r";
+	int err;
+
+	LOG_DBG("");
+
+	eswifi_lock(eswifi);
+
+	err = eswifi_request(eswifi, disconnect, strlen(disconnect),
+			     eswifi->buf, sizeof(eswifi->buf));
+	if (err || !eswifi_is_buf_at_ok(eswifi->buf)) {
+		LOG_ERR("Unable to disconnect network");
+		eswifi_unlock(eswifi);
+		return -EIO;
+	}
+
+	eswifi_unlock(eswifi);
+
+	return 0;
+}
+
+static void eswifi_request_work(struct k_work *item)
+{
+	struct eswifi_dev *eswifi;
+	int err;
+
+	LOG_DBG("");
+
+	eswifi = CONTAINER_OF(item, struct eswifi_dev, request_work);
+
+	switch (eswifi->req) {
+	case ESWIFI_REQ_CONNECT:
+		err = eswifi_connect(eswifi);
+		wifi_mgmt_raise_connect_result_event(eswifi->iface, err);
+		break;
+	case ESWIFI_REQ_DISCONNECT:
+		err = eswifi_disconnect(eswifi);
+		wifi_mgmt_raise_disconnect_result_event(eswifi->iface, err);
+		break;
+	case ESWIFI_REQ_SCAN:
+		eswifi_scan(eswifi);
+		break;
+	case ESWIFI_REQ_NONE:
+	default:
+		break;
+	}
+}
+
+static int eswifi_get_mac_addr(struct eswifi_dev *eswifi, u8_t addr[6])
+{
+	char cmd[] = "Z5\r";
+	int err, i, byte = 0;
+
+	err = eswifi_request(eswifi, cmd, strlen(cmd), eswifi->buf,
+			     sizeof(eswifi->buf));
+	if (err) {
+		return err;
+	}
+
+	for (i = 0; i < sizeof(eswifi->buf); i++) {
+		if (i < 2) {
+			continue;
+		}
+
+		if (eswifi->buf[i] == ':' ||
+		    ((byte == 5) && (eswifi->buf[i] == '\r'))) {
+			addr[byte++] = strtol(&eswifi->buf[i - 2], NULL, 16);
+		}
+	}
+
+	if (byte != 6) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void eswifi_iface_init(struct net_if *iface)
+{
+	struct eswifi_dev *eswifi = &eswifi0;
+	u8_t mac[6];
+	int err;
+
+	LOG_DBG("");
+
+	eswifi_lock(eswifi);
+
+	err = eswifi_reset(eswifi);
+	if (err) {
+		LOG_ERR("Unable to reset device");
+		return;
+	}
+
+	err = eswifi_get_mac_addr(eswifi, mac);
+	if (err) {
+		LOG_ERR("Unable to read MAC address");
+		return;
+	}
+
+	LOG_DBG("MAC Address %02X:%02X:%02X:%02X:%02X:%02X",
+		   mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+
+	net_if_set_link_addr(iface, mac, sizeof(mac), NET_LINK_ETHERNET);
+
+	eswifi->iface = iface;
+
+	eswifi_unlock(eswifi);
+
+	eswifi_offload_init(eswifi);
+}
+
+static int eswifi_mgmt_scan(struct device *dev, scan_result_cb_t cb)
+{
+	struct eswifi_dev *eswifi = dev->driver_data;
+
+	LOG_DBG("");
+
+	eswifi_lock(eswifi);
+	eswifi->scan_cb = cb;
+	eswifi->req = ESWIFI_REQ_SCAN;
+	eswifi_unlock(eswifi);
```

Comment:
Shouldn't this unlock be after the k_work_submit_to_queue()?

Reply:
fixed